server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/dash/sources/coinmarketcap/v2/CoinmarketcapRateSourceV2.java

```diff
@@ -0,0 +1,115 @@
+package com.generalbytes.batm.server.extensions.extra.dash.sources.coinmarketcap.v2;
+
+import com.generalbytes.batm.server.extensions.Currencies;
+import com.generalbytes.batm.server.extensions.IRateSource;
+import si.mazi.rescu.RestProxyFactory;
+
+import java.math.BigDecimal;
+import java.util.*;
+import java.util.concurrent.TimeUnit;
+
+/**
+ *
+ */
+public class CoinmarketcapRateSourceV2 implements IRateSource {
+    /**
+     * Expiry of cache in seconds
+     */
+    private static final long CACHE_EXPIRY_TIME_DEFAULT = 600;
+
+    private static long recentUnix = System.currentTimeMillis();
+
+    private static Map<String,Integer> coinIDs;
+
+    private ICoinmarketcapV2API api;
+
+    public CoinmarketcapRateSourceV2() {
+        api = RestProxyFactory.createProxy(ICoinmarketcapV2API.class, "https://api.coinmarketcap.com");
+        final long currentUnix = System.currentTimeMillis();
+        final long difference = currentUnix - recentUnix;
+        final long differenceInSeconds = TimeUnit.SECONDS.convert(difference, TimeUnit.MILLISECONDS);
+        if(coinIDs == null || coinIDs.isEmpty() || differenceInSeconds > CACHE_EXPIRY_TIME_DEFAULT) {
+            coinIDs = new HashMap<String, Integer>();
```

Comment:
Two problems with this:
1) the static field should be volatile. Otherwise, some threads might not be using the updated value.
2) after a new empty `Map` is assigned to the field, the threads that see the new value will stop getting rates until the map is filled. The solution is to first create a local variable, fill it afterwards and assign it to the filed at the end.